[ES|QL] New @kbn/esql-services package

dej611 opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/172649

This PR extracts the ES|QL service like "validation", "autocomplete", "code_action" (quick fixes) and some other services into a separate package `@kbn/esql-validation-autocomplete` which has no tie with Monaco editor anymore.

All the AST and grammar logic has been encapsulated into a single `@kbn/esql-ast` package who contains the `getAst` function together with the `parser` and `lexer`.

## Validation enhancements

The validation API has been enhanced to gracefully degrade whenever callbacks are not provided, via the `ignoreOnMissingCallback` option. Tests are included as well for this scenario.

### Example plugin

An example plugin app has been developed to show to to use the validation API

<img width="500" alt="Screenshot 2024-03-22 at 17 45 38" src="https://github.com/elastic/kibana/assets/924948/ebc172d3-ee61-4f3a-9e42-dcb9b15c7e69">

The app starts with a missing callback:

<img width="855" alt="Screenshot 2024-03-22 at 17 36 22" src="https://github.com/elastic/kibana/assets/924948/c9b81370-8a10-487d-b22f-2359e1365a54">

Toggling the ignore option will hide the index error due to the lack of `getSources` callback:

<img width="816" alt="Screenshot 2024-03-22 at 17 28 17" src="https://github.com/elastic/kibana/assets/924948/931f4e5f-3ad6-46f7-97ca-63d7bb66646b">

Warnings are shown as well when detected:

<img width="807" alt="Screenshot 2024-03-22 at 17 35 56" src="https://github.com/elastic/kibana/assets/924948/5e3e0537-cba2-475f-946b-0302867384ca">

#### Code snippet

The example app produces a copyable code snippet of the current configuration of the validator:

![Code snippet](https://private-user-images.githubusercontent.com/924948/316838987-fd16da58-8790-40b2-851d-b92aca1811dc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0NTY0MzUsIm5iZiI6MTcxMTQ1NjEzNSwicGF0aCI6Ii85MjQ5NDgvMzE2ODM4OTg3LWZkMTZkYTU4LTg3OTAtNDBiMi04NTFkLWI5MmFjYTE4MTFkYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyNlQxMjI4NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZDhhOGRlMjZjMzZhN2NjNTdkYWMxYTNjOWVkNTA3NDk5NGViOTg2NWEzMTAyZGM4MGYyOTQ0YjQ2N2Q3ZDI1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.H-3r_B57NK1-rS8UbR_bxlNtAMQ0flFGtW0oTIiOlr0)

### Checklist

- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
/ci

elasticmachine commented:
Pinging @elastic/kibana-visualizations (Team:Visualizations)

drewdaemon commented on packages/kbn-esql-services/src/autocomplete/types.ts:
```diff
@@ -0,0 +1,28 @@
+}
+
+export interface EditorContext {
+  /** The actual char that triggered the suggestion (1 single char) */
+  triggerCharacter?: string;
+  /** The type of trigger id. triggerKind = 0 is a programmatic trigger, while any other non-zero value is currently ignored. */
+  triggerKind: number;
```
These comments are helpful. Might be nice to see some on `SuggestionRawDefinition`. For example, what is `kind`? What is `asSnippet`?

dej611 replied:
Made the `kind` prop more explicit about its meaning with 90de4196714c6197ec8cccea694a7f82ce3c1d41

drewdaemon reviewed (commented):
Ran out of time, but will finish reviewing tomorrow. Looking amazing!

stratoula commented on packages/kbn-esql-ast/README.md:
```diff
@@ -0,0 +1,93 @@
+
+To update the grammar:
+1. start by copying the source grammar `lexer` and `parser` files
+2. make sure to fix all the case insensitive occurrencies in `lexer` file (all non-symbol strings like `"something"` into `S O M E T H I N G`).
```
Yes we can remove this step now

dej611 replied:
Resolved with https://github.com/elastic/kibana/commit/c05582f2af261af86cf3c50c16f721e5cf8e2a7e

stratoula commented on packages/kbn-esql-services/README.md:
```diff
@@ -0,0 +1,181 @@
+
+#### Autocomplete
+
+This is the complete logic for the ES|QL autocomplete language, it is completely indipendent from the actual editor (i.e. Monaco) and the suggestions reported need to be wrapped against the specific editor shape.
```
```suggestion
This is the complete logic for the ES|QL autocomplete language, it is completely indepedent from the actual editor (i.e. Monaco) and the suggestions reported need to be wrapped against the specific editor shape.
```

dej611 replied:
Resolved with c05582f2af261af86cf3c50c16f721e5cf8e2a7e

stratoula reviewed (commented):
@dej611 looks fantastic ❤️ 

In the example can we have an EUICode to showcase how we should use the function. So for example if I have only the `getPolicies` callback checked then the api is .... Have the preview on an EUICode component. It will be faster for an engineer to copy paste I think.

dej611 commented:
Added a live codeblock after the Text input component easy to copy and paste ( 9712431884303f580c57ca669e7bb69fdad7f914 )

<img width="730" alt="Screenshot 2024-03-26 at 12 07 55" src="https://github.com/elastic/kibana/assets/924948/1e463ff6-c736-4877-9d77-42ada5a8b6bd">
<img width="725" alt="Screenshot 2024-03-26 at 12 07 40" src="https://github.com/elastic/kibana/assets/924948/fd16da58-8790-40b2-851d-b92aca1811dc">

For the `getFieldsFor` callback I've added some note about the callback argument too.

dej611 commented:
/ci

stratoula reviewed (approved):
Looks awesome @dej611 🎉 Grazie!

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
After offline discussion with @stratoula agreed on renaming the `@kbn/esql-services` package into `@kbn/esql-validation-autocomplete`. This should close #179345 too in one go.

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
Hopefully all renaming issue should be solved at this point 🤞

dej611 commented:
/ci

dej611 commented:
/ci

dej611 commented:
The `SuggestionRawDefinition['kind']` refactor made it pull some `monaco` extra stuff into the `kbn-monaco` package adding a little bit more weight into the shared package: +5.7KB vs +2.6KB before.
If that's ok, then we're good.

dej611 commented:
/ci

dej611 commented:
/ci

drewdaemon commented on packages/kbn-esql-validation-autocomplete/src/shared/context.ts:
```diff
@@ -136,7 +136,20 @@ function isBuiltinFunction(node: ESQLFunction) {
 }
 
-export function getAstContext(innerText: string, ast: ESQLAst, offset: number) {
+/**
+ * Given a ES|QL query string, its AST and the cursor position,
+ * it returns the type of context for the position ("list", "function", "option", "setting", "expression", "newCommand")
+ * plus the whole hierarchy of nodes (command, option, setting and actual position node) context.
+ *
+ * Type details:
+ * * "list": the cursor is inside a "in" list of values (i.e. `a in (1, 2, <here>)`)
+ * * "function": the cursor is inside a function call (i.e. `fn(<here>)`)
+ * * "option": the cursor is inside a command option (i.e. `command ... by <here>`)
+ * * "setting": the cursor is inside a setting (i.e. `command _<here>`)
+ * * "expression": the cursor is inside a command expression (i.e. `command ... <here>` or `command a = ... <here>`)
+ * * "newCommand": the cursor is at the beginning of a new command (i.e. `command1 | command2 | <here>`)
+ */
```
Very helpful comment!

drewdaemon reviewed (approved):
Amazing work. Thank you

stratoula commented:
As there is a conflict can we move these packages to the new @elastic/kibana-esql team?

dej611 commented on packages/kbn-esql-validation-autocomplete/README.md:
```diff
@@ -0,0 +1,191 @@
+
+## Folder structure
+
+This library enables all the advanced features for ES|QL within Monaco, as validation, autocomplete, hover, etc...
```
```suggestion
This library enables all the advanced features for ES|QL, as validation, autocomplete, hover, etc...
```

dej611 commented:
/ci

dej611 commented:
/ci
